### PR TITLE
[Analytics] Save user events to firebase

### PR DESCRIPTION
**Pull request by MarkDavies0**
## Changes
Adds userEvents 

## Requirements

Requires: 
N/A

## PR Dependencies

Depends on: #184 

## Testing

- [ ] Manual

### CODE REVIEW

File: common/database/repositories/UserEventsRepo.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+    async setByUserId(userId: string, state: UserEvents) {
+        const docs = await this.getDocs(userId);
+        const data = { userId, data: state };
+        if (docs.length < 1) {
+            await this.create(data);
+        } else if (docs.length === 1) {
+            await docs[0].ref.set(data);
+        } else {
+            throw new Error('assertion failed: should not have more than one state per user');
```
Comment (wynnset) [inline common/database/repositories/UserEventsRepo.ts:33]: Error message should say user events collection, not user state

File: common/database/repositories/UserEventsRepo.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        }
+    }
+
+    async updateUserEvents(userId: string, state: Partial<UserEvents>): Promise<void | FirebaseFirestore.WriteResult> {
+        const docs = await this.getDocs(userId);
+        if (docs.length === 1) {
+            return await docs[0].ref.set({ data: state }, { merge: true });
+        } else {
+            throw new Error('assertion failed: should have only one state per user');
```
Comment (wynnset) [inline common/database/repositories/UserEventsRepo.ts:42]: Error message should say user events collection, not user state

File: common/database/repositories/UserEventsRepo.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        if (docs.length < 1) {
+            return null;
+        } else {
+            return docs[0];
+        }
+    }
+
+    async setByUserId(userId: string, state: UserEvents) {
+        const docs = await this.getDocs(userId);
+        const data = { userId, data: state };
```
Comment (wynnset) [inline common/database/repositories/UserEventsRepo.ts:27]: `event`, not `state`

File: common/database/repositories/UserEventsRepo.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            await docs[0].ref.set(data);
+        } else {
+            throw new Error('assertion failed: should not have more than one state per user');
+        }
+    }
+
+    async updateUserEvents(userId: string, state: Partial<UserEvents>): Promise<void | FirebaseFirestore.WriteResult> {
+        const docs = await this.getDocs(userId);
+        if (docs.length === 1) {
+            return await docs[0].ref.set({ data: state }, { merge: true });
```
Comment (wynnset) [inline common/database/repositories/UserEventsRepo.ts:40]: `event`, not `state`

File: common/models/UserEvents.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+};
+
+export type UserEvent = DomainEvent | StrategyEvent | ResourceEvent;
+export type UserEvents = Array<UserEvent>;
+
+// - Selected Domains
+// - Selected Strategies
+// - Add a resource to favourites
+// - Remove a resource from favourites
+// - Hide a resource (by pressing “X” and confirming)
```
Comment (wynnset) [inline common/models/UserEvents.ts:28]: Why are these here?

File: mobile/src/controllers/UserEventsController.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    if (!this._current) {
+      // INITIAL STATE for UserEvents after first login by user
+      this._current = [];
+      updateDiff = this._current;
+    }
+
+    this._synced.trigger();
+
+    if (updateDiff) {
+      const needsUserState = await RepoFactory.Instance.userEvents.needsUserEvents(this._uid);
```
Comment (wynnset) [inline mobile/src/controllers/UserEventsController.ts:47]: Should be `needsUserEvents` or something?

File: mobile/src/viewModels/DomainViewModel.ts [hunk 7]
```diff
@@ -2,6 +2,7 @@ import { observable } from 'mobx';
 import { createLogger } from 'common/logger';
 import { Domain, DomainName, DomainSlug, FocusedDomains, Subdomain, SubdomainName } from '../constants/Domain';
 import AppController from 'src/controllers';
+import { UserEvent } from 'common/models/UserEvents';
```
Comment (wynnset) [inline mobile/src/viewModels/DomainViewModel.ts:5]: Is this needed? Doesn't seem to be used

File: mobile/src/stateMachine/views/strategies/ReviewStrategiesView.tsx [hunk 8]
```diff
@@ -97,7 +110,14 @@ export class ReviewStrategiesView extends ViewState {
...
           {this.renderInnerContent('Here are your focus strategies.', true)}
-          {this.showButton && <Button title='CONTINUE' style={styles.selectButton} onPress={() => this.onNext()} theme={this.theme} />}
+          {this.showButton && <Button title='CONTINUE' style={styles.selectButton} disabled={this.buttonsDisabled} onPress={() => this.onNext()} theme={this.theme} />}
+          {this.buttonsDisabled &&
+            <Portal>
+              <View style={{ width: '100%', height: '100%', flexDirection: 'column', justifyContent: 'center', alignItems: 'center' }}>
+                <Text style={[TextStyles.labelMedium, { padding: 20, backgroundColor: 'white', borderRadius: 10, borderWidth: 1, borderColor: 'white' }]}>Loading...</Text>
+              </View>
+            </Portal>
+          }
```
Comment (wynnset) [inline mobile/src/stateMachine/views/strategies/ReviewStrategiesView.tsx:120]: Just making sure... this is not related to user events right?

**Review by wynnset — COMMENTED** [review]
@MarkDavies0 just a few minor things

File: mobile/src/stateMachine/views/strategies/ReviewStrategiesView.tsx (hunk 8 above)
Comment (MarkDavies0) [reply mobile/src/stateMachine/views/strategies/ReviewStrategiesView.tsx:121]: not directly related but still an important change
